api/v1/resourceset_types.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2024 Stefan Prodan.
+// SPDX-License-Identifier: AGPL-3.0
+
+package v1
+
+import (
+	"strings"
+	"time"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/fluxcd/pkg/apis/meta"
+)
+
+const (
+	ResourceSetKind = "ResourceSet"
+)
+
+// ResourceSetSpec defines the desired state of ResourceSet
+type ResourceSetSpec struct {
+	// CommonMetadata specifies the common labels and annotations that are
+	// applied to all resources. Any existing label or annotation will be
+	// overridden if its key matches a common one.
+	// +optional
+	CommonMetadata *CommonMetadata `json:"commonMetadata,omitempty"`
+
+	// Inputs contains the list of resource group inputs.
+	// +optional
+	Inputs []ResourceSetInput `json:"inputs,omitempty"`
+
+	// Resources contains the list of Kubernetes resources to reconcile.
+	// +optional
+	Resources []*apiextensionsv1.JSON `json:"resources,omitempty"`
+
+	// DependsOn specifies the list of Kubernetes resources that must
+	// exist on the cluster before the reconciliation process starts.
+	// +optional
+	DependsOn []Dependency `json:"dependsOn,omitempty"`
+
+	// The name of the Kubernetes service account to impersonate
+	// when reconciling the generated resources.
+	// +optional
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+
+	// Wait instructs the controller to check the health
+	// of all the reconciled resources.
+	// +optional
+	Wait bool `json:"wait,omitempty"`
```

Comment:
The first thing that comes across my mind reading this field is "custom healthchecks".

Reply:
Also there should be a timeout field.

Reply:
The timeout is set via annotations as explained in the docs, this is the convention used by all Flux Operator APIs.